### ISSUE TITLE
[JENKINS-63520] Agent remoting deadlock after reboot

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/remoting#1422]
<p>When we upgrade and reboot the Jenkins agents, sometimes they hang on startup. We have about 50 agents and we upgrade/reboot them twice a day. About 1/100 times an agent will get stuck on startup. </p>

<p>On the Jenkins master, we see this error message from the hung agent's logs:</p>

```
ERROR: Connection terminated
java.nio.channels.ClosedChannelException
	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)
	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:795)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

<p>From the hung agent, we see the attached <tt>jstack</tt> thread dump with deadlock found. It looks like two threads are waiting on each other which causes the deadlock. After encountering this deadlock, the agent never finishes connecting to the master. The master is unable to use the agent as a node when it reaches this hung state.</p>

<p>Could the fact that the java versions are different contribute to this problem? The master has version <tt>1.8.0_252-8u252-b09-1~18.04-b09</tt> whereas the agents have java version <tt>1.8.0_265-8u265-b01-0</tt>. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=anhuong">anhuong</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63520" target="_blank">Agent remoting deadlock after reboot</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jthompson">jthompson</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: remoting
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2023-02-14T18:50:55+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 6
<li><b>imported</b>: 2025-11-25
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When we upgrade and reboot the Jenkins agents, sometimes they hang on startup. We have about 50 agents and we upgrade/reboot them twice a day. About 1/100 times an agent will get stuck on startup. </p>

<p>On the Jenkins master, we see this error message from the hung agent's logs:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>ERROR: Connection terminated
java.nio.channels.ClosedChannelException
	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)
	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:795)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
</pre>
</div></div>

<p>From the hung agent, we see the attached <tt>jstack</tt> thread dump with deadlock found. It looks like two threads are waiting on each other which causes the deadlock. After encountering this deadlock, the agent never finishes connecting to the master. The master is unable to use the agent as a node when it reaches this hung state.</p>

<p>Could the fact that the java versions are different contribute to this problem? The master has version <tt>1.8.0_252-8u252-b09-1~18.04-b09</tt> whereas the agents have java version <tt>1.8.0_265-8u265-b01-0</tt>. </p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins version: 2.235.3<br/>
Remoting version: 4.3<br/>
Java versions:<br/>
&nbsp;&nbsp;Master: java.runtime.version 1.8.0_252-8u252-b09-1~18.04-b09<br/>
&nbsp;&nbsp;Master: java.version 1.8.0_252<br/>
&nbsp;&nbsp;Agent: OpenJDK Runtime Environment (build 1.8.0_265-8u265-b01-0ubuntu2~18.04-b01)
```
</details>
<details><summary><i>1 attachment</i></summary>

- [jstack.txt](https://issues.jenkins.io/secure/attachment/52443/jstack.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bhartshorn.svg" title="bhartshorn's avatar" /> bhartshorn:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63520?focusedId=408947&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-408947">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Jeff &amp; Uli,</p>

<p>Thanks for the fix and testing. I'm part of the team that reported the issue, but we don't have bandwidth to test a custom build at the moment. Is it looking this patch will make it into a release soon?</p>
</pre>
</details>

</ul>
</details>

<p>Hi Jeff & Uli,</p>

<p>Thanks for the fix and testing. I'm part of the team that reported the issue, but we don't have bandwidth to test a custom build at the moment. Is it looking this patch will make it into a release soon?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jthompson:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63520?focusedId=408954&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-408954">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It's expected to be in the next release (4.8) as soon as we can get the release out.</p>
</pre>
</details>

</ul>
</details>

<p>It's expected to be in the next release (4.8) as soon as we can get the release out.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>basil:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63520?focusedId=435183&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-435183">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This was fixed in <a href="https://github.com/jenkinsci/remoting/pull/445" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/remoting/pull/445</a> and the fix has been present in Jenkins weeklies since 2.294 and LTS releases since 2.303.1. If you continue to have issues, please file a new ticket with steps to reproduce the problem from scratch.</p>
</pre>
</details>

</ul>
</details>

<p>This was fixed in <a href="https://github.com/jenkinsci/remoting/pull/445" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/remoting/pull/445</a> and the fix has been present in Jenkins weeklies since 2.294 and LTS releases since 2.303.1. If you continue to have issues, please file a new ticket with steps to reproduce the problem from scratch.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>schwannden:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63520?focusedId=436665&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-436665">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Jenkins 2.346.3 with remoting 4.13.3 is still experiencing the same issue. Agent setup stops in</p>

<p>Aug 27, 2020 8:00:02 AM hudson.remoting.jnlp.Main$CuiListener status INFO: Remote identity confirmed: xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx</p>

<p>And fail to continue connection.</p>

<p> </p>

<p>Env</p>

<p>Jenknis agent setup behind NAT</p>

<p>Jenkins 2.346.3</p>

<p>Remoting: 4.13.3</p>

<p>JRE: openjdk 8</p>
</pre>
</details>

</ul>
</details>

<p>Jenkins 2.346.3 with remoting 4.13.3 is still experiencing the same issue. Agent setup stops in</p>

<p>Aug 27, 2020 8:00:02 AM hudson.remoting.jnlp.Main$CuiListener status INFO: Remote identity confirmed: xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx</p>

<p>And fail to continue connection.</p>

<p> </p>

<p>Env</p>

<p>Jenknis agent setup behind NAT</p>

<p>Jenkins 2.346.3</p>

<p>Remoting: 4.13.3</p>

<p>JRE: openjdk 8</p>
